### PR TITLE
Handling of action (/me) messages

### PR DESCRIPTION
**Pull request by walle303**
Make action message render properly like so:
Example

Also add a /me command for sending action messages.

### CODE REVIEW

**Comment by peppy** [comment]
This does solve the problem, but the code is not at all what we want.

- make an `IsAction` property in `Message`. Use it instead of manual checks everywhere.
- make the returned content from a message never include control characters. move the json input to a private field and make `Content` a property. Also move all of this to a `LegacyMessage : Message` as to not pollute the base class.
- Your code styling doesn't match our project.

File: osu.Game/Overlays/Chat/ChatLine.cs [hunk 1]
```diff
@@ -62,6 +62,42 @@ private Color4 getUsernameColour(Message message)
...
+                return $@"*{message.Sender.Username}";
+            } else {
+                return $@"{message.Sender.Username}:";
+            }
+        }
+
+        private string getMessageDisplayString(Message message)
+        {
+            if (message.Content.StartsWith("\u0001ACTION") && message.Content.EndsWith("\u0001")) {
+                return message.Content.Substring(8,message.Content.Length-9);
```
Comment (peppy) [inline osu.Game/Overlays/Chat/ChatLine.cs]: ugly

File: osu.Game/Overlays/Chat/ChatLine.cs [hunk 2]
```diff
@@ -62,6 +62,42 @@ private Color4 getUsernameColour(Message message)
             return username_colours[message.UserId % username_colours.Length];
         }
 
+        private string getUsernameDisplayString(Message message)
```
Comment (peppy) [inline osu.Game/Overlays/Chat/ChatLine.cs]: all these methods are overkill and shouldn't be required. just inline the changes

**Review by peppy — CHANGES_REQUESTED** [review]
.

**Comment by walle303** [comment]
I think that covers the changes you wanted.

**Comment by peppy** [comment]
It's looking much better already. I'll do another pass once you've had a look at the class changes 👍

**Comment by walle303** [comment]
It might be better to implement the handling of action messages and control characters serverside, so that action messages can be implemented in the new chat system as well.

It appears [ppy/osu-web/app/Http/Controllers/API/ChatController.php](https://github.com/ppy/osu-web/blob/master/app/Http/Controllers/API/ChatController.php) contains the code for this, am I correct?

**Comment by peppy** [comment]
You are correct. We really should be doing this stuff server-side.

**Comment by walle303** [comment]
That should cover it until you go over ppy/osu-web#1201

File: osu.Game/Overlays/ChatOverlay.cs [hunk 3]
```diff
@@ -356,10 +357,24 @@ private void postMessage(TextBox textbox, bool newText)
...
-                // TODO: handle commands
-                currentChannel.AddNewMessages(new ErrorMessage("Chat commands are not supported yet!"));
-                textbox.Text = string.Empty;
-                return;
+                // TODO: Add more commands
+                if (postText.StartsWith("/me "))
+                {
+                    postText = postText.Remove(0,4);
+                    postIsAction = true;
+                } else if (postText == "/me") {
```
Comment (peppy) [inline osu.Game/Overlays/ChatOverlay.cs]: please format your code in line with the rest of the project. we don't ever put `else if` on the same line as braces

File: osu.Game/Overlays/Chat/ChatLine.cs [hunk 4]
```diff
@@ -13,7 +13,8 @@ namespace osu.Game.Overlays.Chat
 {
     public class ChatLine : Container
     {
-        public readonly Message Message;
+
```
Comment (peppy) [inline osu.Game/Overlays/Chat/ChatLine.cs]: unnecessary newline

**Review by peppy — CHANGES_REQUESTED** [review]
fix the few remaining formatting issues

File: osu.Game/Overlays/ChatOverlay.cs (hunk 3 above)
Comment (peppy) [inline osu.Game/Overlays/ChatOverlay.cs]: also all of this code feels wrong. get the command portion of the word and split out the rest before doing any processing, then use `switch` on the command part.

**Comment by peppy** [comment]
For future pull requests please make a branch before PRing (don't use your master branch).

**Comment by walle303** [comment]
Does that take care of the formatting changes?
(I usually do have a separate branch, just forgot to do so this time)

**Comment by peppy** [comment]
Also are you on discord? Would be easier to discuss this implementation if you are.

**Comment by walle303** [comment]
I don't use discord for a variety of reasons and would prefer to use an IRC channel or [matrix](https://matrix.org/) room, but seeing as you already have a fairly large discord room (based on how many people the badge says are in it), that's probably not going to happen. [Well... there is a way to bridge a matrix room and a discord room, but the bridge is very unfinished] If you'd like to contact me via Matrix, my Matrix ID is [@walle303:kek.community](https://matrix.to/#/@walle303:kek.community)

**Comment by peppy** [comment]
Okay so basically, you're adding help messages that aren't what I want to be displayed.

Keep each command in its own `case` block, including any error messages.

**Comment by walle303** [comment]
What would you rather the help messages say, or would you rather change them yourself afterwards?

File: osu.Game/Overlays/ChatOverlay.cs [hunk 5]
```diff
@@ -354,12 +355,34 @@ private void postMessage(TextBox textbox, bool newText)
 
             if (currentChannel == null) return;
 
-            if (postText[0] == '/')
+            switch (postText.Split(' ')[0])
```
Comment (peppy) [inline osu.Game/Overlays/ChatOverlay.cs]: check for the / before entering the switch..

**Comment by peppy** [comment]
Something needs to be done about this:

untitled

**Comment by walle303** [comment]
Is it possible to clear the textbox after postMessage returns? It looks like textbox is part of osu-framework.

File: osu.Game/Overlays/Chat/ChatLine.cs [hunk 6]
```diff
@@ -109,12 +109,13 @@ public ChatLine(Message message)
                 {
                     RelativeSizeAxes = Axes.X,
                     AutoSizeAxes = Axes.Y,
-                    Padding = new MarginPadding { Left = message_padding + padding },
+                    Padding = new MarginPadding { Left = message_padding + (Message.IsAction ? 5 : 15) },
```
Comment (DrabWeb) [inline osu.Game/Overlays/Chat/ChatLine.cs]: `Message.IsAction ? 5 : padding`

File: osu.Game/Online/Chat/Message.cs [hunk 7]
```diff
@@ -46,8 +49,8 @@ public Message(long id)
 
         public bool Equals(Message other) => Id == other?.Id;
 
-        public override int GetHashCode() => Id.GetHashCode();
-    }
+        public override int GetHashCode() => Id.GetHashCode();
```
Comment (peppy) [inline osu.Game/Online/Chat/Message.cs]: what's happening here?

**Comment by peppy** [comment]
This has been implemented server-side. May be best to just rebase this and rewrite as a bit of the underlying code has changed. Let me know if you are still wanting to continue implementing this.

**Comment by walle303** [comment]
Yeah, sorry for the delay, was a bit busy, I can still working on this.

**Comment by peppy** [comment]
`is_action` is now being correctly returned by the server, for what it's worth.

**Comment by peppy** [comment]
Closing due to inactivity. Please reopen when you address the issues.

**Pull request closed without merging**